controlplane/kubeadm/controllers/remediation.go

```diff
@@ -0,0 +1,180 @@
+			}
+		}
+	}()
+
+	// before starting remediation, run preflight checks in order to verify it is safe to remediate.
+	// If those checks are failing, we surface the reason why remediation is not happening into the MachineOwnerRemediated condition
+	// and then we return so KCP can proceed with other operations (ctrl.Result nil).
+
+	var desiredReplicas int = 0
+	if c.KCP.Spec.Replicas != nil {
+		desiredReplicas = int(*c.KCP.Spec.Replicas)
+	}
+
+	// The cluster MUST have spec.replicas >= 3, because this is the smallest cluster size that allows any etcd failure tolerance.
+	if desiredReplicas < 3 {
+		logger.Info("A control plane machine needs remediation, but there are less than 3 control plane machines. Skipping remediation", "UnhealthyMachine", machineToBeRemediated.Name, "Replicas", desiredReplicas)
+		conditions.MarkFalse(machineToBeRemediated, clusterv1.MachineOwnerRemediatedCondition, clusterv1.WaitingForRemediationReason, clusterv1.ConditionSeverityWarning, "KCP can't remediate if there are less than 3 control plane machines")
+		return ctrl.Result{}, nil
+	}
+
+	// The number of replicas MUST be equal to or greater than the desired replicas. This rule ensures that when the cluster
+	// is missing replicas, we skip remediation and instead perform regular scale up/rollout operations first.
+	if c.Machines.Len() < desiredReplicas {
+		logger.Info("A control plane machine needs remediation, but the current number of replicas is currently lower that expected. Skipping remediation", "UnhealthyMachine", machineToBeRemediated.Name, "Replicas", desiredReplicas, "CurrentReplicas", c.Machines.Len())
+		conditions.MarkFalse(machineToBeRemediated, clusterv1.MachineOwnerRemediatedCondition, clusterv1.WaitingForRemediationReason, clusterv1.ConditionSeverityWarning, "KCP waiting for having at least %d control plane machines before triggering remediation", desiredReplicas)
+		return ctrl.Result{}, nil
+	}
+
+	// The cluster MUST have no machines with a deletion timestamp. This rule prevents KCP taking actions while the cluster is in a transitional state.
+	if c.HasDeletingMachine() {
+		logger.Info("A control plane machine needs remediation, but there are other control-plane machines being deleted. Skipping remediation", "UnhealthyMachine", machineToBeRemediated.Name)
+		conditions.MarkFalse(machineToBeRemediated, clusterv1.MachineOwnerRemediatedCondition, clusterv1.WaitingForRemediationReason, clusterv1.ConditionSeverityWarning, "KCP waiting for control plane machine deletion to complete before triggering remediation")
+		return ctrl.Result{}, nil
+	}
+
+	// Remediation MUST preserve etcd quorum. This rule ensures that we will not remove a member that would result in etcd
+	// losing a majority of members and thus become unable to field new requests.
+	if c.IsEtcdManaged() {
+		canSafelyRemediate, err := r.canSafelyRemoveEtcdMember(ctx, c, machineToBeRemediated)
+		if err != nil {
+			return ctrl.Result{}, err
```

Comment:
The issue in `canSafelyRemoveEtcdMember ` may get fixed during `reconcileHealth` so we need to move forward in case of error IMO. WDYT?

Reply:
Actually, just saw that you are already doing this below by returning nil in case `!canSafelyRemediate`. 
How about here though? Not being able to connect to the workload cluster is a reason `canSafelyRemoveEtcdMember ` return error, but this check actually belongs to `reconcileHealth`.

Reply:
I'm not sure I follow your comments.
ReconcileHealth does not apply any change to the cluster, so how can be it fixes an issue in  `canSafelyRemoveEtcdMember`?
Also, how do you envision to consume the output of "Not being able to connect to the workload cluster " from `reconcileHealth` in this func? And how this could prevent this func to connect to the local cluster and handling the error?

Reply:
It is not about fixing, just the correct place to return the connection error. 
`reconcileHealth ` already returns with error if it fails to connect to the workload cluster.
My thinking was not to return with errors as much as possible here if it is not directly related to remediation, but it is just a nit.

Reply:
Thanks for the context.
I got your point, but from the other side we cannot prevent other func to connect to the local cluster and handling the error